include/dmlc/logging.h

```diff
@@ -428,30 +426,42 @@ class LogMessageFatal : public LogMessage {
 #else
 class LogMessageFatal {
  public:
-  LogMessageFatal(const char* file, int line) {
-    log_stream_ << "[" << pretty_date_.HumanDate() << "] " << file << ":"
-                << line << ": ";
+  LogMessageFatal(const char *file, int line) {
+    Entry::ThreadLocal()->Init(file, line);
   }
-  std::ostringstream &stream() { return log_stream_; }
-  ~LogMessageFatal() DMLC_THROW_EXCEPTION {
+  std::ostringstream &stream() { return Entry::ThreadLocal()->log_stream; }
+  DMLC_NO_INLINE ~LogMessageFatal() DMLC_THROW_EXCEPTION {
```

Comment:
The no-inline trick here instructs the compiler not to merge the complicated logic of throwing in the destructor into the caller. This usually wouldn't hurt performance because it is quite a slow path